server/src/main/java/org/elasticsearch/index/codec/DeduplicatingFieldInfosFormat.java

```diff
@@ -0,0 +1,94 @@
+import org.apache.lucene.store.IOContext;
+import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.common.util.StringLiteralDeduplicator;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.index.mapper.FieldMapper;
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * Wrapper around a {@link FieldInfosFormat} that will deduplicate and intern all field names, attribute-keys and -values, and in most
+ * cases attribute maps on read. We use this to reduce the per-field overhead for Elasticsearch instances holding a large number of
+ * segments.
+ */
+public final class DeduplicatingFieldInfosFormat extends FieldInfosFormat {
+
+    private static final Map<Map<String, String>, Map<String, String>> attributeDeduplicator = ConcurrentCollections.newConcurrentMap();
+
+    private static final StringLiteralDeduplicator attributesDeduplicator = new StringLiteralDeduplicator();
+
+    private final FieldInfosFormat delegate;
+
+    public DeduplicatingFieldInfosFormat(FieldInfosFormat delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public FieldInfos read(Directory directory, SegmentInfo segmentInfo, String segmentSuffix, IOContext iocontext) throws IOException {
+        final FieldInfos fieldInfos = delegate.read(directory, segmentInfo, segmentSuffix, iocontext);
+        final FieldInfo[] deduplicated = new FieldInfo[fieldInfos.size()];
+        int i = 0;
+        for (FieldInfo fi : fieldInfos) {
+            deduplicated[i++] = new FieldInfo(
+                FieldMapper.internFieldName(fi.getName()),
+                fi.number,
+                fi.hasVectors(),
+                fi.omitsNorms(),
+                fi.hasPayloads(),
+                fi.getIndexOptions(),
+                fi.getDocValuesType(),
+                fi.getDocValuesGen(),
+                // original attributes is UnmodifiableMap
+                internStringStringMap(fi.attributes()),
+                fi.getPointDimensionCount(),
+                fi.getPointIndexDimensionCount(),
+                fi.getPointNumBytes(),
+                fi.getVectorDimension(),
+                fi.getVectorEncoding(),
+                fi.getVectorSimilarityFunction(),
+                fi.isSoftDeletesField(),
+                fi.isParentField()
+            );
+        }
+        return new FieldInfos(deduplicated);
+    }
+
+    private static Map<String, String> internStringStringMap(Map<String, String> m) {
+        var res = attributeDeduplicator.get(m);
+        if (res == null) {
+            if (attributeDeduplicator.size() > 100) {
+                // Unexpected edge case to have more than 100 different attribute maps
+                // Just to be safe, don't retain more than 100 maps to prevent a potential memory leak
+                attributeDeduplicator.clear();
+            }
+            final Map<String, String> interned = Maps.newHashMapWithExpectedSize(m.size());
```

Comment:
Just a safeguard: can we avoid caching if the size of the attributes is greater than 10?

Reply:
Sure thing :) will do before merging!